projects/plugins/jetpack/extensions/blocks/ai-paragraph/edit.js

```diff
@@ -7,219 +7,272 @@ import { useSelect } from '@wordpress/data';
-	if ( categoryNames.length ) {
-		return sprintf(
-			/** translators: This will be a prompt to OpenAI to generate a post based on the post title, and comma-seperated category names. */
-			__( "This is a post titled '%1$s', published in categories '%2$s':\n", 'jetpack' ), // eslint-disable-line @wordpress/i18n-no-collapsible-whitespace
-			title,
-			categoryNames
-		);
-	}
-
-	// Title and content
-	if ( content ) {
-		return sprintf(
-			/** translators: This will be a prompt to OpenAI to generate a post based on the post title, and the last 240 characters of content. */
-			__( "This is a post titled '%1$s':\n\n…%2$s", 'jetpack' ), // eslint-disable-line @wordpress/i18n-no-collapsible-whitespace
-			title,
-			content
-		);
+			setTimeout( () => setContent( html ), 50 * tokens.length );
+		} else {
+			setContent( html );
+		}
+		setContentSet( true );
 	}
 
-	return sprintf(
-		/** translators: This will be a prompt to OpenAI to generate a post based on the post title */
-		__( 'Write content of a post titled "%s"', 'jetpack' ),
-		title
+	return (
+		<>
+			<div className="content">
+				<RawHTML>{ content }</RawHTML>
+			</div>
+		</>
 	);
 }
 
-/**
- * Gather data available in Gutenberg and prepare the best prompt we can come up with.
- *
- * @param {Function} select - function returning Gutenberg data store.
- * @returns {string} - prompt ready to pipe to OpenAI.
- */
-function preparePromptBasedOnEditorState( select ) {
-	const editorContent = formatEditorCanvasContentForCompletion(
-		select( 'core/block-editor' )
-	).slice( -240 );
+export default function Edit( { attributes, setAttributes, clientId } ) {
+	const [ isLoadingCompletion, setIsLoadingCompletion ] = useState( false );
+	const [ isLoadingCategories, setIsLoadingCategories ] = useState( false );
+	const [ showAnimation, setShowAnimation ] = useState( false );
+	const [ needsMoreCharacters, setNeedsMoreCharacters ] = useState( 0 );
+	const [ showRetry, setShowRetry ] = useState( false );
+	const [ errorMessage, setErrorMessage ] = useState( false );
 
 	// Let's grab post data so that we can do something smart.
-	const currentPost = select( 'core/editor' ).getCurrentPost();
-	if ( ! currentPost ) {
-		return '';
-	}
+	const currentPostTitle = useSelect(
+		select => select( 'core/editor' ).getEditedPostAttribute( 'title' ),
+		[]
+	);
 
-	// If there is no title, there is not much we can do.
-	if ( ! currentPost.title ) {
-		return createPrompt( '', editorContent, '' );
-	}
+	let loading = false;
+	const categories = useSelect(
+		select => select( 'core/editor' ).getEditedPostAttribute( 'categories' ),
+		[]
+	);
 
-	// We are filtering out the default "Uncategorized"
-	const categories = currentPost.categories.filter( catId => catId !== 1 );
-	const tags = currentPost.tags;
+	const categoryObjects = useSelect(
+		select => {
+			return categories
+				.map( categoryId => {
+					const category = select( 'core' ).getEntityRecord( 'taxonomy', 'category', categoryId );
```

Comment:
It would be nice if this somehow was also grabbed from editor cache, but we can solve this later

Reply:
It should be called once, and internally most of it is probably cached anyway...

Reply:
It is not! Inspect the network tab.
But this can be another issue.